torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -590,6 +593,15 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
     torch::lazy::BackendDataPtr handle =
         WrapXlaData(xla::ComputationClient::Get()->CreateDataPlaceholder(
             device.toString(), std::move(shape)));
+    // if SPMD is enabled, we assume all output will be replicated
+    if (ShardingUtil::UseVirtualDevice()) {
+      XLATensor::ShardingSpecPtr sharding =
+          std::make_shared<XLATensor::ShardingSpec>(
+              xla::HloSharding::Replicate().ToProto(), shape);
+      handle = WrapXlaData(xla::ComputationClient::Get()->WrapDataShards(
```

Comment:
It seems that you are already creating a sharded data handle in case of virtual device, maybe we can update the comment, "if SPMD is enabled, we assume all output will be replicated" to "If SPMD is enabled, we assume all output will be sharded or replicated and wrapped inside PjRtShardedData handle."; this change should work either way.

Reply:
sg, let me do that.